### ISSUE TITLE
Enable XRCoordinateSystem to support dynamic coordinate systems

### ISSUE THREAD

**Issue by thetuvix** [immersive-web/webxr#384]
One gotcha in the base API at the moment is that all `XRCoordinateSystem` instances can only represent **logically static objects**, as the `XRCoordinateSystem.getTransformTo(coordinateSystem)` method is not time-indexed.

While the UA can adjust the relationships between two static `XRCoordinateSystem`s over time as the system learns about the world, missing time-indexing means an `XRCoordinateSystem` cannot represent a dynamic object that will physically move between now and a particular photon time. Instead, developers locate dynamic objects like a head or controller by asking a given time's `XRFrame` for an `XRDevicePose` or `XRInputPose`, respectively.

Limiting `XRCoordinateSystem` to static objects has worked OK for the current features in the spec, but it means that, as we build out our AR support, we don't have a durable currency that apps can use to represent a dynamically-moving object, such as the head or controller to track as part of a hit-test subscription.

Note that a common currency to represent a dynamic coordinate system will likely be useful for multiple WebXR APIs we are likely to need over time, for both VR and AR:
* A hit-test subscription method on `XRSession` that takes a base `XRCoordinateSystem` to track each frame (e.g. a device-locked `XRFrameOfReference`), along with an `XRRay` within that coordinate system ([hit-test #4](https://github.com/immersive-web/hit-test/issues/4))
* A coordinate system that tracks the root pose of detected AR world objects, some of which may require time-indexing (e.g. an ARKit ARFaceAnchor, etc.)
* A way to specify the reprojection basis for a second compositor layer to track, e.g. so the compositor can independently stabilize a handheld object to the motion of that hand during reprojection

### Proposal

The core proposal is to upgrade `XRCoordinateSystem` to represent either static or dynamic coordinate systems, by allowing time-indexing when relating two coordinate systems:
* Replace `XRCoordinateSystem.getTransformTo(coordinateSystem)` with **`XRFrame.getPose(coordinateSystem, baseCoordinateSystem)`**, getting the pose of `coordinateSystem` within `baseCoordinateSystem` at the time represented by that `XRFrame`.

This approach continues to support static coordinate systems, while enabling us to then introduce various dynamic coordinate systems over time, as we define VR and AR features that depend on durable tracking of dynamic objects.

Examples of features that will depend on dynamic coordinate systems:
* **`XRSession.subscribeHitTest(XRCoordinateSystem coordinateSystem, XRRay ray)`** to allow for delivery each frame of a hit-test result that aligns with that frame's photon time, solving [hit-test #4](https://github.com/immersive-web/hit-test/issues/4).
* **`XRPlaneAnchor/XRFaceAnchor/XRMeshAnchor.coordinateSystem`** as we introduce detected anchor types for world understanding.
* **`XRLayer.setBasisCoordinateSystem`** to allow the compositor to reproject a layer independently from others, based on a coordinate system that tracks a dynamic object. Examples of basis coordinate systems:
  * **`XRFrameOfReferenceType.device`** (or some explicit subtype of `XRCoordinateSystem`) as a durable coordinate system that can be used for device-locked objects positioned relative to the head/device over time.  This could perhaps replace `XRDevicePose.poseModelMatrix`.
  * **`XRInputSource.gripCoordinateSystem`** as a durable coordinate system that can be used for hand-locked objects positioned relative to the hand/controller over time.  This could perhaps replace `XRInputPose.gripMatrix`.

Thoughts?

**Comment by toji** [comment]
+1 to Blair's sentiment. I'm a big fan of making everything timing-dependent utilize the `XRFrame` object, so this proposal sounds good to me!

Not sure how I feel about the name `getPose` specifically. It wasn't immediately intuitive to me when I first read it, but Alex's following one-line explanation cleared it up entirely. Given the other `get*Pose` functions already on the frame I would suggest `getCoordinateSystemPose` but since that's pretty clunky. I wouldn't object to an interface rename to `XRSpace` (as suggested in [Anchors #4](https://github.com/immersive-web/anchors/issues/4#issuecomment-407914651)) and `getSpacePose` instead.

I've got thoughts about the various examples Alex posted too, but given that they're tangential to the purpose of this issue I'm not gonna dive into those right now.

**Comment by speigg** [comment]
Makes sense to me. Just wondering if there a rationale for `getPose` on the frame instance  

```
XRFrame.getPose(cooridnateSystem, baseCooridnateSystem)
```

Over something like:

```
XRCoordinateSystem.getPose(coordinateSystem, frame)
```

**Comment by toji** [comment]
@speigg: They're both perfectly functional in terms of code structure. I'd say the biggest thing favoring the former is that it's the pattern we've used for every other frame-dependent query in the API thus far. As I pointed out in my previous comment, though, I think simply using `getPose` would be a poor choice for a function on `XRFrame`. It'd be fine for a function on `XRCoordinateSystem` due to clearer context.

**Comment by lincolnfrog** [comment]
This is heavily related to hit-testing in AR. See https://github.com/immersive-web/hit-test/pull/31

**Comment by lincolnfrog** [comment]
I have an IDL proposal based on the hit-test requirements I outlined in immersive-web/hit-test#31:

Either on XRSession:
XRCoordinateSystem getDeviceCoordinateSystem(XRDevice device);
XRCoordinateSystem getDeviceCoordinateSystem(XRInputSource input_source);

or on XRDevice and XRInputSourceDirectly:
XRCoordinateSystem getCoordinateSystem(XRDevice device);

These coordinate systems would be automatically updated to reflect the pose of the XRDevice. This requires the XRCoordinateSystem.getPose(baseCoordinateSystem, frame) option above.

**Comment by NellWaliczek** [comment]
Fixed by #460
